src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,506 @@
+   * the larger odd value for half should be returned, or the smaller even
+   * number.
+   * @param weightOdd
+   * @private
+   */
+  private getHalfOfSlideCount(weightOdd: boolean): number {
+    const halfLength = (this.getSlides().length - 1) / 2;
+    if (halfLength % 2 === 0) {
+      return halfLength;
+    } else if (weightOdd) {
+      return Math.ceil(halfLength);
+    } else {
+      return Math.floor(halfLength);
+    }
+  }
+
+  /**
+   * Return half of the slides by taking slides to one side of the given slide.
+   *
+   * This function assumes looping is enabled and loops around if needed to get
+   * half of the slides.
+   */
+  private splitSlidesInHalf(
+      slide: HTMLElement, direction: Direction
+  ): HTMLElement[] {
+    const targetLength =
+        this.getHalfOfSlideCount(direction === Direction.RIGHT);
+    const result = [];
+    let indexToAdd = this.getSlideIndex(slide);
+    while (result.length < targetLength) {
+      // Build the looped index
+      indexToAdd = mathf.wrap(indexToAdd + direction, 0, this.getSlideCount());
+      const slideToAdd = this.getSlideByIndex(indexToAdd);
+      if (direction === Direction.RIGHT) {
+        result.push(slideToAdd);
+      } else {
+        result.unshift(slideToAdd);
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Setup initial values.
+   */
+  private init(): void {
+    this.transition.init(this.getFirstSlide(), this);
+    this.raf.start();
+  }
+
+  /**
+   * Run the loop to do all the necessary work for the carousel.
+   */
+  private loop(): void {
+    this.raf.read(() => {
+      if (!this.condition()) {
+        return;
+      }
+
+      const shouldSync: boolean = this.shouldSync();
+      const activeSlide = this.getActiveSlide();
+      if (activeSlide !== this.lastActiveSlide) {
+        this.lastActiveSlide = activeSlide;
+        this.onTransitionCallbacks.forEach((callback) => callback(this));
+
+        if (activeSlide) {
+          this.raf.write(() => this.updateClasses(activeSlide));
+        }
+
+        // Sync other carousels.
+        if (shouldSync) {
+          this.synchronizer
+              .handleCarouselTransition(this, this.getSlideIndex(activeSlide));
+        }
+      }
+    });
+  }
+
+  /**
+   * Update the slide elements' CSS classes
+   * @param activeSlide
+   */
+  private updateClasses(activeSlide: HTMLElement) {
+    const slidesBefore = this.getSlidesBefore(activeSlide);
+    const slidesAfter = this.getSlidesAfter(activeSlide);
+
+    const adjustCssClasses =
+        (slide: HTMLElement, cssClassesToKeep: Set<string>) => {
+          const currentCssClasses = this.slideCssClasses.get(slide);
+          this.slideCssClasses.set(slide, cssClassesToKeep);
+          const classesToRemove =
+              setf.subtract(currentCssClasses, cssClassesToKeep);
+          slide.classList.remove(...classesToRemove);
+          slide.classList.add(...cssClassesToKeep);
```

Comment:
Maybe I'm misunderstanding what these lines do, but instead of allocating sets and calculating differences, couldn't you do something like:

```
el.className = classesToKeep; // as string
```

Reply:
This would prevent other code from being able to work with the classlist of the same element.

Reply:
The extra tracking of class names in sets in this class is so that the carousel only removes and adds its own classes. If we weren't concerned with other potential code modifying the element we could use the `el.classList` features instead but we risk creating side effects.

Reply:
Resolved with attribute setup suggested in other comment.